diff --git a/lib/librumpuser/Makefile b/lib/librumpuser/Makefile
index 2994dc8..df6ce82 100644
--- a/lib/librumpuser/Makefile
+++ b/lib/librumpuser/Makefile
@@ -35,19 +35,19 @@ SRCS+=		rumpuser_sp.c
 .error Unsupported curlwp scheme for thread model: ${RUMP_CURLWP}
 .endif
 .endif
-SRCS=		rumpfiber.c rumpfiber_bio.c
+SRCS=		rumpfiber.c
 SRCS+=		rumpfiber_sp.c
 .else
 .error Unsupported rumpuser threading type: ${RUMPUSER_THREADS}
 .endif
 
 SRCS+=		rumpuser_component.c rumpuser_random.c
-SRCS+=		rumpuser_file.c rumpuser_mem.c
+SRCS+=		rumpuser_file_dummy.c rumpuser_mem.c
 
 SRCS+=		rumpuser_errtrans.c rumpuser_sigtrans.c
 
 # optional
-SRCS+=		rumpuser_dl.c rumpuser_daemonize.c
+SRCS+=		rumpuser_dl.c rumpuser_daemonize_dummy.c
 
 INCSDIR=	/usr/include/rump
 INCS=		rumpuser_component.h rumpuser_port.h
diff --git a/lib/librumpuser/rumpfiber.c b/lib/librumpuser/rumpfiber.c
index 4a19383..7d879fd 100644
--- a/lib/librumpuser/rumpfiber.c
+++ b/lib/librumpuser/rumpfiber.c
@@ -71,7 +71,6 @@
 __RCSID("$NetBSD: rumpfiber.c,v 1.9 2014/12/29 21:50:09 justin Exp $");
 #endif /* !lint */
 
-#include <sys/ioctl.h>
 #include <sys/mman.h>
 #include <sys/time.h>
 
@@ -398,6 +397,9 @@ init_sched(void)
 {
 	struct thread *thread = calloc(1, sizeof(struct thread));
 
+	if (! thread)
+		abort();
+
 	thread->name = strdup("init");
 	thread->flags = 0;
 	thread->wakeup_time = -1;
@@ -482,7 +484,7 @@ int
 rumpuser_clock_sleep(int enum_rumpclock, int64_t sec, long nsec)
 {
 	enum rumpclock rclk = enum_rumpclock;
-	uint32_t msec;
+	uint64_t msec;
 	int nlocks;
 
 	rumpkern_unsched(&nlocks, NULL);
@@ -511,14 +513,17 @@ rumpuser_getparam(const char *name, void *buf, size_t blen)
 		strncpy(buf, ncpu, blen);
 		rv = 0;
 	} else if (strcmp(name, RUMPUSER_PARAM_HOSTNAME) == 0) {
-		char tmp[MAXHOSTNAMELEN];
-
-		if (gethostname(tmp, sizeof(tmp)) == -1) {
-			snprintf(buf, blen, "rump-%05d", (int)getpid());
-		} else {
-			snprintf(buf, blen, "rump-%05d.%s",
-			    (int)getpid(), tmp);
-		}
+		strncpy(buf, "rump", blen);
+	//} else if (strcmp(name, RUMPUSER_PARAM_HOSTNAME) == 0) {
+		//char tmp[MAXHOSTNAMELEN];
+
+		//if (gethostname(tmp, sizeof(tmp)) == -1) {
+		//	snprintf(buf, blen, "rump-%05d", (int)getpid());
+		//} else {
+		//	snprintf(buf, blen, "rump-%05d.%s",
+		//	    (int)getpid(), tmp);
+		//}
+	//	strncpy(buf, rump, blen);
 		rv = 0;
 	} else if (*name == '_') {
 		rv = EINVAL;
@@ -562,11 +567,12 @@ rumpuser_seterrno(int error)
 void
 rumpuser_dprintf(const char *format, ...)
 {
-	va_list ap;
+	//va_list ap;
 
-	va_start(ap, format);
-	vfprintf(stderr, format, ap);
-	va_end(ap);
+	printk("something\n");
+	//va_start(ap, format);
+	//vfprintf(stderr, format, ap);
+	//va_end(ap);
 }
 
 int
@@ -682,8 +688,8 @@ rumpuser_mutex_init(struct rumpuser_mtx **mtxp, int flags)
 {
 	struct rumpuser_mtx *mtx;
 
-	mtx = malloc(sizeof(*mtx));
-	memset(mtx, 0, sizeof(*mtx));
+	mtx = calloc(1, sizeof(*mtx));
+	if (! mtx) abort();
 	mtx->flags = flags;
 	TAILQ_INIT(&mtx->waiters);
 	*mtxp = mtx;
@@ -766,8 +772,8 @@ rumpuser_rw_init(struct rumpuser_rw **rwp)
 {
 	struct rumpuser_rw *rw;
 
-	rw = malloc(sizeof(*rw));
-	memset(rw, 0, sizeof(*rw));
+	rw = calloc(1, sizeof(*rw));
+	if (!rw) abort();
 	TAILQ_INIT(&rw->rwait);
 	TAILQ_INIT(&rw->wwait);
 
@@ -900,8 +906,8 @@ rumpuser_cv_init(struct rumpuser_cv **cvp)
 {
 	struct rumpuser_cv *cv;
 
-	cv = malloc(sizeof(*cv));
-	memset(cv, 0, sizeof(*cv));
+	cv = calloc(1, sizeof(*cv));
+	if (!cv) abort();
 	TAILQ_INIT(&cv->waiters);
 	*cvp = cv;
 }
diff --git a/lib/librumpuser/rumpfiber.h b/lib/librumpuser/rumpfiber.h
index aefbfac..4aed7be 100644
--- a/lib/librumpuser/rumpfiber.h
+++ b/lib/librumpuser/rumpfiber.h
@@ -40,7 +40,7 @@ struct thread {
     int64_t wakeup_time;
     TAILQ_ENTRY(thread) thread_list;
     ucontext_t ctx;
-    uint32_t flags;
+    int flags;
     int threrrno;
 };
 
diff --git a/lib/librumpuser/rumpfiber_sp.c b/lib/librumpuser/rumpfiber_sp.c
index e96cbdb..d8183e5 100644
--- a/lib/librumpuser/rumpfiber_sp.c
+++ b/lib/librumpuser/rumpfiber_sp.c
@@ -33,6 +33,7 @@
 __RCSID("$NetBSD: rumpfiber_sp.c,v 1.3 2014/12/29 21:50:09 justin Exp $");
 #endif /* !lint */
 
+#include <stdint.h>
 #include <stdlib.h>
 
 #include <rump/rumpuser.h>
diff --git a/lib/librumpuser/rumpuser_daemonize_dummy.c b/lib/librumpuser/rumpuser_daemonize_dummy.c
new file mode 100644
index 0000000..671712a
--- /dev/null
+++ b/lib/librumpuser/rumpuser_daemonize_dummy.c
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2014 Justin Cormack.  All Rights Reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS
+ * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include "rumpuser_port.h"
+
+#if !defined(lint)
+__RCSID("$NetBSD$");
+#endif /* !lint */
+
+#include <stdint.h>
+#include <stdlib.h>
+
+#include "rumpuser_int.h"
+
+#include <rump/rumpuser.h>
+
+int
+rumpuser_daemonize_begin(void)
+{
+
+	abort();
+}
+
+int
+rumpuser_daemonize_done(int error)
+{
+
+	abort();
+}
diff --git a/lib/librumpuser/rumpuser_file_dummy.c b/lib/librumpuser/rumpuser_file_dummy.c
new file mode 100644
index 0000000..88f03cd
--- /dev/null
+++ b/lib/librumpuser/rumpuser_file_dummy.c
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2007-2010 Antti Kantee.  All Rights Reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS
+ * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+/* NOTE this code will move to a new driver in the next hypercall revision */
+
+#include "rumpuser_port.h"
+
+#if !defined(lint)
+__RCSID("$NetBSD: rumpuser_file.c,v 1.1 2014/07/09 23:41:40 justin Exp $");
+#endif /* !lint */
+
+#include <errno.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include <rump/rumpuser.h>
+
+#include "rumpuser_int.h"
+
+int
+rumpuser_getfileinfo(const char *path, uint64_t *sizep, int *ftp)
+{
+
+	ET(ENOENT);
+}
+
+int
+rumpuser_open(const char *path, int ruflags, int *fdp)
+{
+
+	ET(ENOENT);
+}
+
+int
+rumpuser_close(int fd)
+{
+
+	ET(EBADF);
+}
+
+int
+rumpuser_iovread(int fd, struct rumpuser_iovec *ruiov, size_t iovlen,
+	int64_t roff, size_t *retp)
+{
+
+	ET(EBADF);
+}
+
+int
+rumpuser_iovwrite(int fd, const struct rumpuser_iovec *ruiov, size_t iovlen,
+	int64_t roff, size_t *retp)
+{
+
+	ET(EBADF);
+}
+
+int
+rumpuser_syncfd(int fd, int flags, uint64_t start, uint64_t len)
+{
+
+	ET(EBADF);
+}
+
+void
+rumpuser_bio(int fd, int op, void *data, size_t dlen, int64_t doff,
+        rump_biodone_fn biodone, void *bioarg)
+{
+
+	biodone(bioarg, (size_t)0, EBADF); /* convert error no */
+}
diff --git a/lib/librumpuser/rumpuser_mem.c b/lib/librumpuser/rumpuser_mem.c
index 880640e..0b1d5e0 100644
--- a/lib/librumpuser/rumpuser_mem.c
+++ b/lib/librumpuser/rumpuser_mem.c
@@ -47,24 +47,40 @@ int
 rumpuser_malloc(size_t howmuch, int alignment, void **memp)
 {
 	void *mem = NULL;
-	int rv;
+/* use mmap if alignment possible */
+#ifdef MAP_ALIGNED
+	int af = 0;
 
+	while (alignment > 1) {
+		alignment >>= 1;
+		af++;
+	}
+
+	mem = mmap(0, howmuch, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANON | MAP_ALIGNED(af), -1, 0);
+	if (__predict_false(mem == MAP_FAILED)) {
+		ET(EINVAL);
+	}
+#else
 	if (alignment == 0)
 		alignment = sizeof(void *);
 
-	rv = posix_memalign(&mem, (size_t)alignment, howmuch);
-	if (__predict_false(rv != 0)) {
-		if (rv == EINVAL) {
-			printf("rumpuser_malloc: invalid alignment %d\n",
-			    alignment);
-			abort();
-		}
+	mem = aligned_alloc((size_t)alignment, howmuch);
+	if (__predict_false(mem == NULL)) {
+		ET(EINVAL);
 	}
-
+#endif
 	*memp = mem;
-	ET(rv);
+	ET(0);
 }
 
+#ifdef MAP_ALIGNED
+void
+rumpuser_free(void *ptr, size_t size)
+{
+
+	munmap(ptr, size);
+}
+#else
 /*ARGSUSED1*/
 void
 rumpuser_free(void *ptr, size_t size)
@@ -72,6 +88,7 @@ rumpuser_free(void *ptr, size_t size)
 
 	free(ptr);
 }
+#endif
 
 int
 rumpuser_anonmmap(void *prefaddr, size_t size, int alignbit,
@@ -83,8 +100,9 @@ rumpuser_anonmmap(void *prefaddr, size_t size, int alignbit,
 #ifndef MAP_ALIGNED
 #define MAP_ALIGNED(a) 0
 	if (alignbit)
-		fprintf(stderr, "rumpuser_anonmmap: warning, requested "
-		    "alignment not supported by hypervisor\n");
+/*		fprintf(stderr, "rumpuser_anonmmap: warning, requested "
+		    "alignment not supported by hypervisor\n");*/
+abort();
 #endif
 
 	prot = PROT_READ|PROT_WRITE;
diff --git a/lib/librumpuser/rumpuser_port.h b/lib/librumpuser/rumpuser_port.h
index 82f4361..e141609 100644
--- a/lib/librumpuser/rumpuser_port.h
+++ b/lib/librumpuser/rumpuser_port.h
@@ -174,10 +174,9 @@ getenv_r(const char *name, char *buf, size_t buflen)
 	}
 }
 #endif
-
+/*
 #if !defined(HAVE_POSIX_MEMALIGN)
 #if !defined(HAVE_MEMALIGN)
-#error method for aligned memory allocation required
 #endif
 #include <sys/sysmacros.h>
 #include <stdlib.h>
@@ -191,7 +190,7 @@ posix_memalign(void **ptr, size_t align, size_t size)
 	return 0;
 }
 #endif
-
+*/
 /*
  * For NetBSD, use COHERENCY_UNIT as the lock alignment size.
  * On other platforms, just guess it to be 64.
diff --git a/lib/librumpuser/rumpuser_random.c b/lib/librumpuser/rumpuser_random.c
index 292e921..4494d10 100644
--- a/lib/librumpuser/rumpuser_random.c
+++ b/lib/librumpuser/rumpuser_random.c
@@ -42,13 +42,17 @@ __RCSID("$NetBSD: rumpuser_random.c,v 1.4 2014/11/04 19:05:17 pooka Exp $");
 #include <string.h>
 #include <unistd.h>
 
+#ifdef HAVE_GETRANDOM
+#include <linux/random.h>
+#endif
+
 #include <rump/rumpuser.h>
 
 #include "rumpuser_int.h"
 
 static const size_t random_maxread = 32;
 
-#ifdef HAVE_ARC4RANDOM_BUF
+#if defined(HAVE_ARC4RANDOM_BUF) || defined(HAVE_GETRANDOM)
 int
 rumpuser__random_init(void)
 {
@@ -75,18 +79,28 @@ rumpuser__random_init(void)
 int
 rumpuser_getrandom(void *buf, size_t buflen, int flags, size_t *retp)
 {
-#ifndef HAVE_ARC4RANDOM_BUF
-	ssize_t rv;
+#ifdef HAVE_ARC4RANDOM_BUF
+	buflen = buflen > random_maxread ? random_maxread : buflen;
+	arc4random_buf(buf, buflen);
+	*retp = buflen;
+#elif HAVE_GETRANDOM
+	int rv;
 
-	rv = read(random_fd, buf, buflen > random_maxread ? random_maxread : buflen);
-	if (rv < 0) {
+	buflen = buflen > random_maxread ? random_maxread : buflen;
+	rv = getrandom(buf, buflen, 0);
+	if (rv == -1) {
 		ET(errno);
 	}
 	*retp = rv;
 #else
+	ssize_t rv;
+
 	buflen = buflen > random_maxread ? random_maxread : buflen;
-	arc4random_buf(buf, buflen);
-	*retp = buflen;
+	rv = read(random_fd, buf, buflen);
+	if (rv == -1) {
+		ET(errno);
+	}
+	*retp = rv;
 #endif
 
 	return 0;
